### PR TITLE
feat(behavior_path_planner): output shorten path if ego vehicle is out of route

### PR DESCRIPTION
**Pull request by rej55**
## Description

Before this change, behavior_path_planner outputs an improper path if the ego vehicle is out of route.
To fix it, I add the feature to output shorten path such a situation.

## Tests performed

Tested in planning simulator.

https://user-images.githubusercontent.com/10190493/226199820-d6693759-2bf2-485a-8a82-9a10dab81a36.mp4

TIER IV internal test result 882/882

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_planner/src/behavior_tree_manager.cpp [hunk 1]
```diff
@@ -27,6 +27,43 @@
...
+
+  lanelet::ConstLanelet goal_lane;
+  const bool is_failed_getting_lanelet = std::invoke([&]() {
+    if (isInLanelets(shoulder_lanes, goal_pose)) {
+      return !lanelet::utils::query::getClosestLanelet(shoulder_lanes, goal_pose, &goal_lane);
+    }
+    return !route_handler->getGoalLanelet(&goal_lane);
+  });
+  if (is_failed_getting_lanelet) {
+    PathWithLaneId path{};
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/behavior_tree_manager.cpp:55]: the empty path may cause the node to die.
it's better to create the path from self_pose and offset pose like https://github.com/autowarefoundation/autoware.universe/blob/main/planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp#L541-L572.
how about moving this function to utils and using it?

File: planning/behavior_path_planner/src/behavior_tree_manager.cpp [hunk 2]
```diff
@@ -146,6 +191,51 @@ void BehaviorTreeManager::resetGrootMonitor()
...
+    const auto goal_arc_coord =
+      lanelet::utils::getArcCoordinates({goal_lane}, data->route_handler->getGoalPose());
+    if (ego_arc_coord.length > goal_arc_coord.length + buffer) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  // If ego vehicle is out of the closest lanelet, return false
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/behavior_tree_manager.cpp:226]:
```suggestion
  // If ego vehicle is out of the closest lanelet, return true
```

return true?

File: planning/behavior_path_planner/src/planner_manager.cpp [hunk 3]
```diff
@@ -26,6 +26,42 @@
...
+{
+  for (const auto & lane : lanes) {
+    if (lanelet::utils::isInLanelet(pose, lane)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+PathWithLaneId createGoalAroundPath(
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/planner_manager.cpp:39]: Is it difficult to make these functions common to the one of `behavior_tree_manager.cpp`?

File: planning/behavior_path_planner/src/planner_manager.cpp [hunk 4]
```diff
@@ -416,6 +460,51 @@ void PlannerManager::resetRootLanelet(const std::shared_ptr<PlannerData> & data)
...
+    const auto goal_arc_coord =
+      lanelet::utils::getArcCoordinates({goal_lane}, data->route_handler->getGoalPose());
+    if (ego_arc_coord.length > goal_arc_coord.length + buffer) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  // If ego vehicle is out of the closest lanelet, return false
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/planner_manager.cpp:495]:
```suggestion
  // If ego vehicle is out of the closest lanelet, return true
```

return true?

File: planning/behavior_path_planner/src/behavior_tree_manager.cpp (hunk 1 above)
Comment (kosuke55) [reply planning/behavior_path_planner/src/behavior_tree_manager.cpp:55]: ↑ this function is for shoulder_lanes, so little tedious to make common.

File: planning/behavior_path_planner/src/behavior_tree_manager.cpp (hunk 2 above)
Comment (rej55) [reply planning/behavior_path_planner/src/behavior_tree_manager.cpp]: Fixed!
https://github.com/autowarefoundation/autoware.universe/pull/3075/commits/5557bec97237162a5998d809966373ff1e85f232

File: planning/behavior_path_planner/src/planner_manager.cpp (hunk 3 above)
Comment (rej55) [reply planning/behavior_path_planner/src/planner_manager.cpp]: Fixed!
https://github.com/autowarefoundation/autoware.universe/pull/3075/commits/5557bec97237162a5998d809966373ff1e85f232

File: planning/behavior_path_planner/src/planner_manager.cpp (hunk 4 above)
Comment (rej55) [reply planning/behavior_path_planner/src/planner_manager.cpp]: Fixed!
https://github.com/autowarefoundation/autoware.universe/pull/3075/commits/5557bec97237162a5998d809966373ff1e85f232

File: planning/behavior_path_planner/src/behavior_tree_manager.cpp (hunk 1 above)
Comment (kosuke55) [reply planning/behavior_path_planner/src/behavior_tree_manager.cpp]: It seems ok for me!

**Review by kosuke55 — APPROVED** [review]
LGTM

**Comment by kosuke55** [comment]
FYI:
```
/__w/autoware.universe/autoware.universe/planning/behavior_path_planner/src/utils/utils.cpp: In lambda function:
/__w/autoware.universe/autoware.universe/planning/behavior_path_planner/src/utils/utils.cpp:933:9: error: ‘isInLanelets’ was not declared in this scope; did you mean ‘expandLanelets’?
  933 |     if (isInLanelets(shoulder_lanes, goal_pose)) {
      |         ^~~~~~~~~~~~
      |         expandLanelets
gmake[2]: *** [CMakeFiles/behavior_path_planner_node.dir/build.make:286: CMakeFiles/behavior_path_planner_node.dir/src/utils/utils.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:184: CMakeFiles/behavior_path_planner_node.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
---
  Build behavior_path_planner with OLD framework...


/__w/autoware.universe/autoware.universe/planning/behavior_path_planner/src/utils/utils.cpp: In lambda function:
/__w/autoware.universe/autoware.universe/planning/behavior_path_planner/src/utils/utils.cpp:933:9: error: ‘isInLanelets’ was not declared in this scope; did you mean ‘expandLanelets’?
  933 |     if (isInLanelets(shoulder_lanes, goal_pose)) {
      |         ^~~~~~~~~~~~
      |         expandLanelets
gmake[2]: *** [CMakeFiles/behavior_path_planner_node.dir/build.make:286: CMakeFiles/behavior_path_planner_node.dir/src/utils/utils.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:184: CMakeFiles/behavior_path_planner_node.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
---
Failed   <<< behavior_path_planner [8min 30s, exited with code 2]

Summary: 42 packages finished [35min 20s]
```

**Comment by rej55** [comment]
@kosuke55 Thanks for your report.
I fixed in following:
https://github.com/autowarefoundation/autoware.universe/pull/3075/commits/2fd8df7e1816393f49f92a61dfa6ecb98c67b3da

**Pull request merged**